src/bp/nlu-core/training-worker-queue.ts

```diff
@@ -0,0 +1,253 @@
+        }
+        if (msg.type === 'training_error') {
+          process.off('message', handler)
+          reject(new Error(msg.payload.error!))
+        }
+        if (msg.type === 'training_canceled' && msg.srcWid === workerId) {
+          process.off('message', handler)
+          reject(new TrainingCanceledError())
+        }
+        if (msg.type === 'training_progress') {
+          progress(msg.payload.progress!)
+        }
+      }
+      process.send!(msg)
+      process.on('message', handler)
+    })
+  }
+
+  private _createNewWorker(): Promise<number> {
+    const { config } = this
+    const msg: OutgoingMessage = { type: 'make_new_worker', payload: { config } }
+
+    return new Promise(resolve => {
+      const handler = (msg: IncomingMessage) => {
+        if (msg.type === 'log') {
+          this._logMessage(msg)
+        }
+
+        if (msg.type === 'worker_ready') {
+          process.off('message', handler)
+          resolve(msg.srcWid)
+        }
+      }
+      process.send!(msg)
+      process.on('message', handler)
+    })
+  }
+
+  private _logMessage(msg: IncomingMessage) {
+    const log: Log = msg.payload.log!
+    log.info && this.logger.info(log.info)
+    log.warning && this.logger.warning(log.warning)
+    log.error && this.logger.error(log.error)
+  }
+}
+
+if (cluster.isMaster) {
+  function sendToWebWorker(msg: IncomingMessage) {
+    const webWorker = cluster.workers[process.WEB_WORKER]
+    webWorker?.isConnected() && webWorker.send(msg)
+  }
+
+  function sendToTrainingWorker(msg: OutgoingMessage) {
+    const worker = cluster.workers[msg.destWid!]
+    worker?.send(msg) // TODO: find out why this is sometimes undefined.
+  }
+
+  function killTrainingWorker(msg: OutgoingMessage) {
+    const worker = cluster.workers[msg.destWid!]
+
+    if (!worker) {
+      const response: IncomingMessage = { type: 'training_canceled', payload: {}, srcWid: msg.destWid! }
+      return sendToWebWorker(response)
+    }
+
+    worker!.kill('SIGKILL')
+    const exitHandler = (worker: Worker, _exitCode: number, _signal: string) => {
+      const response: IncomingMessage = { type: 'training_canceled', payload: {}, srcWid: worker.id }
+      sendToWebWorker(response)
+    }
+    cluster.once('exit', exitHandler)
+  }
+
+  registerMsgHandler('make_new_worker', async (msg: OutgoingMessage) => spawnNewTrainingWorker(msg.payload.config!))
+  registerMsgHandler('cancel_training', killTrainingWorker)
+  registerMsgHandler('start_training', sendToTrainingWorker)
+
+  registerMsgHandler('log', sendToWebWorker)
+  registerMsgHandler('worker_ready', sendToWebWorker)
+  registerMsgHandler('training_done', sendToWebWorker)
+  registerMsgHandler('training_progress', sendToWebWorker)
+  registerMsgHandler('training_error', sendToWebWorker)
+}
+
+if (cluster.isWorker && process.env.WORKER_TYPE === WORKER_TYPES.TRAINING) {
+  const config = JSON.parse(process.env.NLU_CONFIG!)
+
+  const srcWid = cluster.worker.id
```

Comment:
explicit variables names won't kill you, we have auto-completion for this : `srcWorkerId` is the shortest you can go